From c67479ab3720ae101331b03e072ae7100c7d7fa0 Mon Sep 17 00:00:00 2001
From: Mohan Prabu <mprabub@qti.qualcomm.com>
Date: Tue, 16 Sep 2025 12:50:17 +0530
Subject: [PATCH] nl80211: Retry cfg80211_get_bss() with NULL SSID on failure

For hidden SSID, kernel might not have scan entry with SSID for
the particular link and on connect notify driver calls
cfg80211_get_bss() with SSID which may result in NULL bss.

Call cfg80211_get_bss() with NULL SSID incase of failure to get bss
as a fallback mechanism to avoid NULL bss for hidden SSID APs.

Change-Id: I7b05ad8370189304e313464aa0321fd73fa07440
Signed-off-by: Mohan Prabu <mprabub@qti.qualcomm.com>
---
 net/wireless/sme.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/net/wireless/sme.c b/net/wireless/sme.c
index cfcc5b1fd666..85e42f21ca63 100644
--- a/net/wireless/sme.c
+++ b/net/wireless/sme.c
@@ -837,6 +837,20 @@ void __cfg80211_connect_result(struct net_device *dev,
 						 wdev->u.client.ssid_len,
 						 wdev->conn_bss_type,
 						 IEEE80211_PRIVACY_ANY);
+			/* In Hidden SSID cases, AP might be stored without
+			 * SSID information in scan results.
+			 * cfg80211_get_bss will return NULL in such cases.
+			 * Retry cfg80211_get_bss() with NULL SSID
+			 * as a fallback.
+			 */
+			if (!cr->links[link].bss) {
+				cr->links[link].bss =
+					cfg80211_get_bss(wdev->wiphy, NULL,
+					cr->links[link].bssid,
+					NULL, 0, wdev->conn_bss_type,
+					IEEE80211_PRIVACY_ANY);
+			}
+
 			if (!cr->links[link].bss) {
 				bss_not_found = true;
 				break;
-- 
2.34.1

